Add operation count based functionality to benchmarks

alanfx opened the pull request:
This PR adds the ability to execute a specific number of operations
during a benchmark.

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyTestStage.java:
```diff
@@ -169,7 +169,18 @@ public void run() {
+      while (!isStarted()) {
+         int stressorStartedCount = 0;
+         for (LegacyStressor stressor : stressors) {
+            if (stressor.isStarted()) {
```
Ouch, busy wait? `LegacyStressor.started` isn't volatile. Why don't you just use `CountDownLatch`?

Btw. the field is not volatile for a purpose, because it's repeatedly read in `recoding()` and given that it was set only once, it's better to keep the result in local field; otherwise we could use stage.isStarted() all the time.

alanfx replied:
I can look into using `CountDownLatch` instead here.

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyStressor.java:
```diff
@@ -87,15 +91,18 @@ private void runInternal() {
-               logic.run(operation);
-            } catch (OperationLogic.RequestException e) {
-               // the exception was already logged in makeRequest
+            } else {
```
This seems like you do one operation, call `stats.begin()` (before the other threads have started) and then just busy loops polling stage.isStarted() without running any ops. I don't think this is what you want.

alanfx replied:
That is exactly what the code does to guarantee that every thread spawned from the stage is ready to run before starting to collect statistics. I thought this was what we discussed?

rvansa replied:
I'll reformulate the problems:
1) Until all other threads are started (`stage.isStarted()` is true), the thread should execute operations in a loop (though these are not recorded). Your change executes only single operation before that.
2) `stats.begin()` should be called as late as possible, before starting to execute recorded operations. Your change calls that right after first operation, before the other threads start (`stage.isStarted()` returns true).

alanfx replied:
Hey @rvansa, I have made the changes, but I am still curious why it matters that the threads execute operations that won't be recorded while the stage is starting up. What is the purpose of that behavior?

rvansa replied:
Because when one thread starts a request, there should be already some other overlapping request being executed - otherwise the request would be uncontended (and likely faster).

alanfx replied:
OK, that makes sense to me. Thanks for explaining this.

alanfx commented:
Updated `LegacyTestStage.startStressors` to use a `CountDownLatch` and moved the call to `stats.begin()` right before the call to `completion.start()` in `LegacyStressor.runInternal`, and I think it is a lot simpler and closer to the ideal. I also set the stage `started` to true after the `rampUp` block. Let me know if you think that's correct.

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyTestStage.java:
```diff
@@ -201,6 +206,14 @@ public void run() {
 
    protected Completion createCompletion() {
       Completion completion = new TimeStressorCompletion(duration);
+      if (numOperations > 0) {
```
Please put the creation of `TimeStressorCompletion` to else branch (you can omit the variable and return new instance right away).

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/TimeStressorCompletion.java:
```diff
@@ -12,6 +12,7 @@
  */
 public class TimeStressorCompletion extends AbstractCompletion {
 
+   protected static final String PROGRESS_STRING = "Number of operations executed by this thread: %d. Elapsed time: %s. Remaining: %s. Total: %s.";
```
Why have you moved this out of `AbstractCompletion`?

alanfx replied:
Because this `PROGRESS_STRING` reports the time elapsed and duration while the one in `CountStressorCompletion` reports executed and total operation count.

rvansa replied:
Ok, my bad.

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyTestStage.java:
```diff
@@ -166,7 +169,9 @@ public void run() {
             throw new IllegalStateException("Interrupted during ramp-up.", e);
          }
       }
+      
```
It's strange that checkstyle does not complain about whitespaces at end of line...

alanfx replied:
No clue here

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyStressor.java:
```diff
@@ -50,10 +53,15 @@ public LegacyStressor(LegacyTestStage stage, OperationLogic logic, int globalThr
+      return this.started && stage.started && !stage.isFinished();
+   }
+
+   public boolean isStarted() {
```
This method is not needed anymore.

rvansa commented on core/src/main/java/org/radargun/stages/test/legacy/LegacyStressor.java:
```diff
@@ -87,18 +95,22 @@ private void runInternal() {
+            } else {
+               if (!isStarted()) {
+                  threadCountDown.countDown();
+                  this.started = true;
```
Okay, this looks much better. The only thing is that I would set `this.started = true` only when we see that `stage.isStarted()` - put that next to `stats.begin()`. For the countdown check, you could use just a local var.
Then, in `recording()` you won't have to check the `stage.started` which is a volatile read.

alanfx commented:
Updated

rvansa commented:
Thanks Alan, integrating.
